pkg/syncer/cnsoperator/controller/cnsregistervolume/util.go

```diff
@@ -95,25 +104,53 @@ func constructCreateSpecForInstance(r *ReconcileCnsRegisterVolume, instance *cns
 }
 
 // getK8sStorageClassName gets the storage class name in K8S mapping the vsphere
-// storagepolicy id.
-func getK8sStorageClassName(ctx context.Context, k8sClient clientset.Interface, storagePolicyID string) (string, error) {
+// storagepolicy id. The policy must also be assigned to the passed namespace.
+func getK8sStorageClassName(ctx context.Context, k8sClient clientset.Interface, storagePolicyID string, namespace string) (string, error) {
 	log := logger.GetLogger(ctx)
 	scList, err := k8sClient.StorageV1().StorageClasses().List(ctx, metav1.ListOptions{})
 	if err != nil {
 		msg := fmt.Sprintf("Failed to get Storageclasses from API server. Error: %+v", err)
 		log.Error(msg)
 		return "", errors.New(msg)
 	}
+	var scName string
 	for _, sc := range scList.Items {
 		scParams := sc.Parameters
 		for paramName, val := range scParams {
 			param := strings.ToLower(paramName)
 			if param == common.AttributeStoragePolicyID && val == storagePolicyID {
-				return sc.Name, nil
+				scName = sc.Name
+				break //There will be only one storage class in the cluster with a given policy ID.
```

Comment:
There can be multiple SC will same storage policy id. 
For example, create a storage policy name `gold` and assign it to SV namespace, it will create a `gold` storage class and create a resource quota for it in that namespace. Later on VC, storage policy is renamed to `gold123`. Now if this storage policy is assigned to SV namespace, it will create `gold123` storage class in SV. This way it is possible that you might multiple storage classes with same policy ID.

Reply:
I think what we can do is to identify all SC with that policy and check if any one of them is assigned to that namespace where `CnsRegisterVolume` is created. If yes, use that SC for PV/PVC creation.

Reply:
Assigning a renamed policy to a SV namespace doesn't create a new storage class in the cluster. I believe WCP fixed that behavior.

Reply:
@gohilankit can you also check this.

create a storage policy name `gold` and assign it to SV namespace, it will create a `gold` storage class and create a resource quota for it in that namespace.

Unassign `gold` policy from that namespace. Later on VC, storage policy is renamed to `gold123`. Now if this storage policy is assigned to SV namespace, would it create `gold123` SC?

Reply:
No it doesn't create a new `gold123` class. The cluster would still have the storage class `gold`.

So the behavior in WCP is - whatever is the name of the storage policy when it first gets assigned to a SV namespace, k8s will have a storage class by that name. Later edits to the policy name and reassignment doesn't have any effect in terms of creating a new SC.

We can make the solution more robust by looping through all k8s SC, but if there are lots of storage classes in the cluster, this will be a performance hit.